src/trace/implementations/merge_batcher.rs

```diff
@@ -173,88 +230,66 @@ impl<D: Ord, T: Ord, R: Semigroup> MergeSorter<D, T, R> {
         }
     }
 
+    /// Helper to get pre-sized vector from the stash.
     #[inline]
-    fn new(logger: Option<Logger<DifferentialEvent, WorkerIdentifier>>, operator_id: usize) -> Self {
-        Self {
-            logger,
-            operator_id,
-            queue: Vec::new(),
-            stash: Vec::new(),
-        }
+    fn empty(&self, stash: &mut Vec<Vec<T>>) -> Vec<T> {
+        stash.pop().unwrap_or_else(|| Vec::with_capacity(self.preferred_buffer_size()))
     }
 
+    /// Helper to return a batch to the stash.
     #[inline]
-    pub fn empty(&mut self) -> Vec<(D, T, R)> {
-        self.stash.pop().unwrap_or_else(|| Vec::with_capacity(Self::buffer_size()))
+    fn recycle(&self, mut batch: Vec<T>, stash: &mut Vec<Vec<T>>) {
+        // TODO: Should we limit the size of `stash`?
+        if batch.capacity() == self.preferred_buffer_size() /*&& stash.len() < 2*/ {
+            batch.clear();
+            stash.push(batch);
+        }
     }
+}
 
-    #[inline]
-    pub fn push(&mut self, batch: &mut Vec<(D, T, R)>) {
-        // TODO: Reason about possible unbounded stash growth. How to / should we return them?
-        // TODO: Reason about mis-sized vectors, from deserialized data; should probably drop.
-        let mut batch = if self.stash.len() > 2 {
-            ::std::mem::replace(batch, self.stash.pop().unwrap())
-        }
-        else {
-            ::std::mem::take(batch)
-        };
+impl<K: Data, V: Data, T: Ord + PartialOrder + Clone+'static,R: Semigroup+'static> Merger for VecMerger<((K,V),T,R)> {
+    type Time = T;
+    type Input = Vec<((K,V),T,R)>;
+    type Batch = Vec<((K,V),T,R)>;
+    type Output = ((K,V), T, R);
 
-        if !batch.is_empty() {
-            crate::consolidation::consolidate_updates(&mut batch);
-            self.account([batch.len()], 1);
-            self.queue_push(vec![batch]);
-            while self.queue.len() > 1 && (self.queue[self.queue.len()-1].len() >= self.queue[self.queue.len()-2].len() / 2) {
-                let list1 = self.queue_pop().unwrap();
-                let list2 = self.queue_pop().unwrap();
-                let merged = self.merge_by(list1, list2);
-                self.queue_push(merged);
+    fn accept(&mut self, batch: RefOrMut<Self::Batch>, stash: &mut Vec<Self::Batch>) -> Vec<Self::Batch> {
+        // `batch` is either a shared reference or an owned allocations.
+        let mut owned = match batch {
+            RefOrMut::Ref(vec) => {
+                let mut owned = self.empty(stash);
+                owned.clone_from(vec);
+                owned
             }
+            RefOrMut::Mut(vec) => std::mem::take(vec)
+        };
+        consolidate_updates(&mut owned);
```

Comment:
Double check whether we should be collecting small batches rather than consolidating and shipping them. The concern is that once they are presented outward to the batcher it will merge batches without thinking about the number of records in them. We want to protect ourselves against quadratic behavior with lots of small elements.